contracts/RenExBalances.sol

```diff
@@ -7,157 +7,107 @@ import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
-        require(traderBalances[_trader][_token] >= _fee, "insufficient funds for fee");
+    /// @notice Transfer a token value from one trader to another, transferring
+    /// a fee to the RewardVault. Can only be called by the RenExSettlement
+    /// contract.
+    ///
+    /// @param _traderFrom The address of the trader to decrement the balance of.
+    /// @param _traderTo The address of the trader to increment the balance of.
+    /// @param _token The token's address.
+    /// @param _value The number of tokens to decrement the balance by (in the
+    ///        token's smallest unit).
+    /// @param _fee The fee amount to forward on to the RewardVault.
+    /// @param _feePayee The recipient of the fee.
+    function transferBalanceWithFee(address _traderFrom, address _traderTo, address _token, uint256 _value, uint256 _fee, address _feePayee)
+    external onlyRenExSettlementContract {
+        require(traderBalances[_traderFrom][_token] >= _fee, "insufficient funds for fee");
 
         if (address(_token) == ETHEREUM) {
-            rewardVaultContract.deposit.value(_fee)(feePayee, ERC20(_token), _fee);
+            rewardVaultContract.deposit.value(_fee)(_feePayee, ERC20(_token), _fee);
         } else {
             ERC20(_token).approve(rewardVaultContract, _fee);
-            rewardVaultContract.deposit(feePayee, ERC20(_token), _fee);
+            rewardVaultContract.deposit(_feePayee, ERC20(_token), _fee);
+        }
+        privateDecrementBalance(_traderFrom, ERC20(_token), _value + _fee);
+        if (_value > 0) {
+            privateIncrementBalance(_traderTo, ERC20(_token), _value);
         }
-        privateDecrementBalance(_trader, ERC20(_token), _value + _fee);
-    }
-
-    function privateDecrementBalance(address _trader, ERC20 _token, uint256 _value) private {
-        require(traderBalances[_trader][_token] >= _value, "insufficient funds");
-        traderBalances[_trader][_token] = traderBalances[_trader][_token].sub(_value);
-
-        emit BalanceDecreased(_trader, _token, _value);
     }
 
-
-    /********** TRADER FUNCTIONS **********************************************/
-
-    /**
-    @notice Deposits ETH or an ERC20 token into the contract
-    @param _token the token's address (must be a registered token)
-    @param _value the amount to deposit in the token's smallest unit
-    */
-    function deposit(ERC20 _token, uint256 _value) payable public {
+    /// @notice Deposits ETH or an ERC20 token into the contract.
+    ///
+    /// @param _token The token's address (must be a registered token).
+    /// @param _value The amount to deposit in the token's smallest unit.
+    function deposit(ERC20 _token, uint256 _value) external payable {
         address trader = msg.sender;
 
         if (address(_token) == ETHEREUM) {
             require(msg.value == _value, "mismatched value parameter and tx value");
         } else {
+            require(msg.value == 0, "unexpected ether transfer");
```

Comment:
To save gas, should we not check this? If someone transfers ETH erroneously that is on them.

Reply:
I think the benefit of this check outweighs the gas. Without the check, the transferred ETH can't be recovered.

Reply:
Fair point, and if it is run by traders then the gas is not as much of an issue.